--- mcclintock/PoPoolationTE/identify-te-insertsites.pl	2015-08-05 16:54:57.000000000 +0100
+++ ../mcclintock/PoPoolationTE/identify-te-insertsites.pl	2015-09-12 11:40:11.000000000 +0100
@@ -36,6 +36,8 @@
     "te-hierarchy-file=s"   =>\$tehierfile,
     "te-hierarchy-level=s"  =>\$tehiertargetlevel,
     "test"                  =>\$test,
+    "insert-distance=i"	    =>\$insertdistance,
+    "read-length=i"	    =>\$readlength,
     "help"	            =>\$help
 ) or pod2usage(-msg=>"Wrong options",-verbose=>1);
 
@@ -415,4 +417,4 @@
 
 =head1 DETAILS
 
-=cut
\ No newline at end of file
+=cut
