--- TEInsert.pm	2012-01-27 09:30:15.000000000 +0000
+++ TEInsert.pm.fixed	2015-03-21 18:44:05.000000000 +0000
@@ -45,13 +45,13 @@
             $popfreq=($fpopfreq+$rpopfreq)/2 if(defined($fpopfreq) and defined($rpopfreq));
             $sitesupport="FR";
         }
-        elsif($frstart)
+        elsif(defined($frstart))
         {
             $popfreq=$fpopfreq;
             $sitesupport="F";
             
         }
-        elsif($rrstart)
+        elsif(defined($rrstart))
         {
             $popfreq=$rpopfreq;
             $sitesupport="R";
@@ -224,10 +224,11 @@
         }
         else
         {
-            die "impossible";
+		return undef;
+            #die "impossible";
         }
         
     }
     
 }
-1;
\ No newline at end of file
+1;
