--- TEInsertUtility.pm	2015-03-19 21:56:19.803561825 +0000
+++ TEInsertUtility.pm.new	2015-03-19 21:55:37.862766962 +0000
@@ -77,7 +77,7 @@
     $roverlap eq "-"and $roverlap=undef;
     
     die "Invalid TE Insertion $l\n" unless $chr;
-    die "Invalid TE Insertion $l\n" unless $inspos;
+    die "Invalid TE Insertion $l\n" unless defined($inspos);
     die "Invalid TE Insertion $l\n" unless $sitesupport;
     die "Invalid TE Insertion $l\n" unless $teid;
     die "Invalid TE Insertion $l\n" unless ($frstart or $rrstart);
