--- TEInsertUtility.pm	2015-03-20 21:14:20.087554546 +0000
+++ TEInsertUtility.pm.new	2015-03-20 21:15:02.950995463 +0000
@@ -78,7 +78,7 @@
     $roverlap eq "-"and $roverlap=undef;
     
     die "Invalid TE Insertion $l\n" unless $chr;
-    die "Invalid TE Insertion $l\n" unless $inspos;
+    die "Invalid TE Insertion $l\n" unless ($inspos || $inspos==0);
     die "Invalid TE Insertion $l\n" unless $sitesupport;
     die "Invalid TE Insertion $l\n" unless $teid;
     die "Invalid TE Insertion $l\n" unless ($frstart or $rrstart);
