--- TEInsert.pm	2012-01-27 09:30:15.000000000 +0000
+++ TEInsert.pm.fixed	2015-03-20 23:30:30.912995797 +0000
@@ -45,13 +45,13 @@
             $popfreq=($fpopfreq+$rpopfreq)/2 if(defined($fpopfreq) and defined($rpopfreq));
             $sitesupport="FR";
         }
-        elsif($frstart)
+        elsif(defined($frstart))
         {
             $popfreq=$fpopfreq;
             $sitesupport="F";
             
         }
-        elsif($rrstart)
+        elsif(defined($rrstart))
         {
             $popfreq=$rpopfreq;
             $sitesupport="R";
@@ -230,4 +230,4 @@
     }
     
 }
-1;
\ No newline at end of file
+1;
