--- TEInsertUtility.pm	2012-01-27 09:30:15.000000000 +0000
+++ TEInsertUtility.pm.fixed	2015-03-21 10:40:36.000000000 +0000
@@ -78,10 +78,10 @@
     $roverlap eq "-"and $roverlap=undef;
     
     die "Invalid TE Insertion $l\n" unless $chr;
-    die "Invalid TE Insertion $l\n" unless $inspos;
+    die "Invalid TE Insertion $l\n" unless ($inspos || $inspos==0);
     die "Invalid TE Insertion $l\n" unless $sitesupport;
     die "Invalid TE Insertion $l\n" unless $teid;
-    die "Invalid TE Insertion $l\n" unless ($frstart or $rrstart);
+    die "Invalid TE Insertion $l\n" unless (defined($frstart) or defined($rrstart));
     
     return TEInsert->new($chr,$inspos,$teid,$order,$fbid,$comment, $frstart,$frend,$fpres,$fabs,$foverlap,$rrstart,$rrend,$rpres,$rabs,$roverlap);
     # INT: chr, inspos, teid, order, fbid, comment, frstart, frend, fpres, fabs, foverlap, rrstart, rrend, rpres, rabs, roverlap
