--- TEMP_Absence.sh	2015-09-09 09:59:07.000000000 +0100
+++ TEMP_Absence_fixed.sh	2015-09-09 10:02:04.000000000 +0100
@@ -116,6 +116,11 @@
 
 rm temp1.sam $i.unpair.sam
 
+# Sometimes $i.unproper.uniq.interval.bed contains malformed bed entries 
+# These must be removed to prevent the script failing
+awk '{if ($3>=$2 && $3 > 0 && $2 > 0) print $0}' $i.unproper.uniq.interval.bed > $i.unproper.uniq.interval.fixed.bed
+mv $i.unproper.uniq.interval.fixed.bed $i.unproper.uniq.interval.bed
+
 # Map to transposons
 bedtools intersect -a $TEBED -b $i.unproper.uniq.interval.bed -f 1.0 -wo > temp
 perl $BINDIR/filterFalsePositive.ex.pl temp $INSERT $i.final.pairs.rpmk.bed
